tool/tsh/db.go

```diff
@@ -858,54 +858,132 @@ func newDatabaseInfo(cf *CLIConf, tc *client.TeleportClient, route tlsca.RouteTo
 	if dbInfo.ServiceName == "" {
 		return nil, trace.BadParameter("missing database service name")
 	}
-	if dbInfo.Protocol == "" {
-		db, err := dbInfo.GetDatabase(cf, tc)
+	if dbInfo.Protocol != "" && dbInfo.Username != "" && dbInfo.Database != "" {
+		return &dbInfo, nil
+	}
+	db, err := dbInfo.GetDatabase(cf, tc)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	dbInfo.Protocol = db.GetProtocol()
+	// If database has admin user defined, we're most likely using automatic
+	// user provisioning so default to Teleport username unless database
+	// username was provided explicitly.
+	if dbInfo.Username == "" && db.GetAdminUser() != "" {
+		log.Debugf("Defaulting to Teleport username %q as database username.", tc.Username)
+		dbInfo.Username = tc.Username
+	}
+	// recheck to see if we can avoid fetching the roleset to set defaults.
+	needDBUser := dbInfo.Username == "" && role.RequireDatabaseUserMatcher(dbInfo.Protocol)
+	needDBName := dbInfo.Database == "" && role.RequireDatabaseNameMatcher(dbInfo.Protocol)
+	if !needDBUser && !needDBName {
+		return &dbInfo, nil
+	}
+
+	profile, err := tc.ProfileStatus()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var proxy *client.ProxyClient
+	err = client.RetryWithRelogin(cf.Context, tc, func() error {
+		proxy, err = tc.ConnectToProxy(cf.Context)
+		return trace.Wrap(err)
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer proxy.Close()
```

Comment:
`tc.ConnectToProxy(cf.Context)` is heavy call and in our cloud env takes ~500ms We should avoid invoking  `ConnectToProxy`from from random places.  We already calling ConnectToProxy from ` dbInfo.GetDatabase` so we should be more  aware about pref impact.

We probably we should take a look at this flow and pref implication before v14 release. 
Could you add a TODO/Reminder comment issue that  we should unify `tc.ConnectToProxy` call in this flow ?

Reply:
I was aware of this extra connecttoproxy call, but I couldn't see a straightforward way to avoid it. Maybe we need to add a way to have TeleportClient reuse a proxy connection instead of opening/closing it for various API calls. I'll make an issue